File: Problem17_SubarraySumEqualsK.java
// Time Complexity : O(n) -> n: length of the array
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public int subarraySum(int[] nums, int k) {
        if(nums == null || nums.length == 0)
            return 0;
        HashMap<Integer, Integer> map = new HashMap<>();
        map.put(0, 1);
        int count = 0, rSum = 0;
        for(int i = 0; i < nums.length; i++) {
            rSum += nums[i];
            if(map.containsKey(rSum - k)) {
                count += map.get(rSum - k);
            } 
            if(!map.containsKey(rSum)) {
                map.put(rSum, 1);
            } else {
                map.put(rSum, map.get(rSum) + 1);
            }
        }
        return count;
    }
}

File: Problem 17 Subarray Sum Equals K.py
# Did this code successfully run on Leetcode: Yes
# Time Complexity: O(n)
# Space Complexity: O(n) since mapper takes cumulative sum & sum can have n values.
# Logic: start calculating the cumulative sum, if sum -k exist in dic that means k sum just passed & increment the count
# with the occurance of sum -k, update the occurance of sum for future sum -k use.

class Solution:
    def subarraySum(self, nums, k):
        mapper = {0: 1}
        count = 0
        sum = 0
        for i in range(len(nums)):
            sum += nums[i]
            if sum - k in mapper:
                count += mapper[sum - k]
            if sum in mapper:
                mapper[sum] += 1
            else:
                mapper[sum] = 1
        return count
